### PR TITLE
MBS-8806: Don’t cache non-existent ACs as empty

### PR DESCRIPTION
**Pull request by uklauer**
When `Data::ArtistCredit::get_by_ids` was called with IDs that didn’t actually exist, it returned empty `ArtistCredit` entities for them. Those were then automatically put into the cache and persisted there. When an artist credit with the relevant ID later actually was created, the empty artist credit was displayed in its place until the old cache entry expired (MBS-8806). It is unclear how `get_by_ids` could be called with a not-yet-existing ID.

Only create and return an `ArtistCredit` entity for IDs actually found in the database. Also emit a stack trace when a non-existent ID is encountered in order to aid further debugging.

### CODE REVIEW

File: lib/MusicBrainz/Server/Data/ArtistCredit.pm [hunk 1]
```diff
@@ -25,26 +27,31 @@ sub get_by_ids
...
-        $counts{$id} = 0;
-    }
     for my $row (@{ $self->sql->select_list_of_hashes($query, @ids) }) {
-        my %info = (
-            artist_id => $row->{artist},
-            name => $row->{ac_name}
-        );
-        $info{join_phrase} = $row->{join_phrase} // '';
-        my $obj = MusicBrainz::Server::Entity::ArtistCreditName->new(%info);
-        $obj->artist($self->c->model('Artist')->_new_from_row($row));
```
Comment (uklauer) [inline lib/MusicBrainz/Server/Data/ArtistCredit.pm]: Unrelated, but I didn’t like this part with the generic `$obj` name and the three different ways to set properties of the new ACN entity, so I rewrote it. (It’s a separate commit if you want to look at the two changes separately.)

File: lib/MusicBrainz/Server/Data/ArtistCredit.pm [hunk 2]
```diff
@@ -25,26 +27,31 @@ sub get_by_ids
                 "JOIN artist ON artist.id=artist_credit_name.artist " .
                 "WHERE artist_credit IN (" . placeholders(@ids) . ") " .
                 "ORDER BY artist_credit, position";
+
     my %result;
     my %counts;
-    foreach my $id (@ids) {
-        my $obj = MusicBrainz::Server::Entity::ArtistCredit->new(id => $id);
-        $result{$id} = $obj;
```
Comment (uklauer) [inline lib/MusicBrainz/Server/Data/ArtistCredit.pm]: This was the problematic part: For every input ID, an output entity was created that remained empty if the database didn’t return anything about it. The `around` modifier in `Role::EntityCache` then cached these.

**Comment by mwiencek** [comment]
Great, thanks 👍

File: lib/MusicBrainz/Server/Data/ArtistCredit.pm [hunk 3]
```diff
@@ -25,26 +27,31 @@ sub get_by_ids
...
-            artist_id => $row->{artist},
-            name => $row->{ac_name}
-        );
-        $info{join_phrase} = $row->{join_phrase} // '';
-        my $obj = MusicBrainz::Server::Entity::ArtistCreditName->new(%info);
-        $obj->artist($self->c->model('Artist')->_new_from_row($row));
         my $id = $row->{artist_credit};
-        $result{$id}->add_name($obj);
+        $counts{$id} //= 0;
+        $result{$id} //= MusicBrainz::Server::Entity::ArtistCredit->new(id => $id);
```
Comment (uklauer) [inline lib/MusicBrainz/Server/Data/ArtistCredit.pm]: This is now inside the loop processing the database results, so the object is only created if the AC actually exists in the database.

File: lib/MusicBrainz/Server/Data/ArtistCredit.pm [hunk 4]
```diff
@@ -25,26 +27,31 @@ sub get_by_ids
...
+        $result{$id}->add_name($acn);
         $counts{$id} += 1;
     }
     foreach my $id (@ids) {
+        if (!defined $counts{$id}) {
+            # It's unclear how this could happen, but it seems to be the
+            # cause of MBS-8806. Perhaps we'll find out with the call trace
+            # provided by cluck.
+            cluck "Attempted to load non-existent AC $id, please investigate";
+            next;
```
Comment (uklauer) [inline lib/MusicBrainz/Server/Data/ArtistCredit.pm]: `cluck` will not die, it only outputs a trace, so `next` explicitly skips to the next iteration.

I really have no idea how it is possible that the method is called with a soon-to-be-but-not-yet ID. It somehow must leak from the transaction creating a new AC, before it commits, but I don’t see how.

**Pull request merged**